sdk/keyvault/azure-keyvault-keys/azure/keyvault/keys/crypto/_models.py

```diff
@@ -2,11 +2,176 @@
+from cryptography.hazmat.primitives.asymmetric.rsa import (
+    RSAPrivateKey,
+    RSAPrivateNumbers,
+    RSAPublicKey,
+    RSAPublicNumbers,
+)
+from cryptography.hazmat.primitives.asymmetric.utils import Prehashed
+from cryptography.hazmat.primitives.hashes import Hash, HashAlgorithm, SHA1, SHA256, SHA384, SHA512
+from cryptography.hazmat.primitives.serialization import Encoding, KeySerializationEncryption, PrivateFormat
+
+from ._enums import EncryptionAlgorithm, KeyWrapAlgorithm, SignatureAlgorithm
 
 if TYPE_CHECKING:
-    from . import EncryptionAlgorithm, KeyWrapAlgorithm, SignatureAlgorithm
-    from typing import Any, Optional
+    # Import clients only during TYPE_CHECKING to avoid circular dependency
+    from ._client import CryptographyClient
+    from .._client import KeyClient
+
+
+SIGN_ALGORITHM_MAP = {
+    SHA256: SignatureAlgorithm.rs256,
+    SHA384: SignatureAlgorithm.rs384,
+    SHA512: SignatureAlgorithm.rs512,
+}
+OAEP_MAP = {
+    SHA1: EncryptionAlgorithm.rsa_oaep,
+    SHA256: EncryptionAlgorithm.rsa_oaep_256
+}
+PSS_MAP = {
+    SignatureAlgorithm.rs256: SignatureAlgorithm.ps256,
+    SignatureAlgorithm.rs384: SignatureAlgorithm.ps384,
+    SignatureAlgorithm.rs512: SignatureAlgorithm.ps512,
+}
+
+
+class ManagedRsaKey(RSAPrivateKey):
+    """An `RSAPrivateKey` implementation based on a key managed by Key Vault.
+
+    :param str key_name: The name of the key vault key to use.
+    :param key_client: The client that will be used to communicate with Key Vault.
+    :type key_client: KeyClient
+    """
+
+    def __init__(self, key_name: str, key_client: "KeyClient") -> None:  # pylint:disable=unused-argument
+        self._key_name: str = key_name
+        self._key_client: "KeyClient" = key_client
+        self._crypto_client: "CryptographyClient" = key_client.get_cryptography_client(key_name)
+
+    def decrypt(self, ciphertext: bytes, padding: AsymmetricPadding) -> bytes:
+        """Decrypts the provided ciphertext.
+
+        :param bytes ciphertext: Encrypted bytes to decrypt.
+        :param padding: The padding to use. Supported paddings are `OAEP` and `PKCS1v15`. For `OAEP` padding, supported
+            hash algorithms are `SHA1` and `SHA256`. The only supported mask generation function is `MGF1`. See
+            https://learn.microsoft.com/azure/key-vault/keys/about-keys-details for details.
+        :type padding: AsymmetricPadding
+
+        :returns: The decrypted plaintext, as bytes.
+        :rtype: bytes
+        """
+        if isinstance(padding, OAEP):
+            # Public algorithm property was only added in https://github.com/pyca/cryptography/pull/9582
+            # _algorithm property has been available in every version of the OAEP class, so we use it as a backup
+            try:
+                algorithm = padding.algorithm
+            except AttributeError:
+                algorithm = padding._algorithm
```

Comment:
Are we happy that these private attributes existed in all the supported versions of cryptography? (Just wondering if we need a final `except AttributeError` here, or if this is safe as far back as our min supported version)

Reply:
Yep! Looking at the [source file](https://github.com/pyca/cryptography/blame/main/src/cryptography/hazmat/primitives/asymmetric/padding.py)'s git blame, `OAEP`'s private attributes have been untouched since the padding was added in https://github.com/pyca/cryptography/pull/949.
Similarly, `MGF1`'s private attributes have been untouched since https://github.com/pyca/cryptography/pull/775, and `PSS`'s since https://github.com/pyca/cryptography/pull/792. These all predate our earliest supported version 🙂